lib/ansible/modules/system/mdadm.py

```diff
@@ -0,0 +1,487 @@
+                           r'(spares=(?P<spares>[\d]+)\s+)?'
+                           r'name=(?P<host>[\w|\.|-]+):(?P<name>[\w|\-|_]+)\s+'
+                           r'UUID=(?P<uuid>[\w+|:]+)\n\s+'
+                           r'devices=(?P<devices>[\/|\w|\-|_|,]+)', re.MULTILINE)
+
+    # list all array detected by mdadm
+    mdadm_cmd = ansmod.get_bin_path("mdadm", required=True)
+    rc, stdout, err = ansmod.run_command(
+        "%s --examine --brief --scan --verbose" % mdadm_cmd,
+        check_rc=True
+    )
+
+    mdarray_list = MDArrayDict()
+    for match in re_examine.finditer(stdout):
+        devices = []
+        for device in match.group('devices').split(','):
+            devices.append(device)
+
+        mdarray = MDArray(
+            name=match.group('name'),
+            dev=match.group('dev'),
+            uuid=match.group('uuid'),
+            level=match.group('level'),
+            metadata=match.group('metadata'),
+            devices=devices,
+            num_devices=match.group('numdevices'),
+            ansmod=ansmod)
+        mdarray_list[mdarray.uuid] = mdarray
+
+    # List all active array
+    mdadm_cmd = ansmod.get_bin_path("mdadm", required=True)
+    rc, stdout, err = ansmod.run_command(
+        "%s --detail --brief --scan --verbose" % mdadm_cmd,
+        check_rc=True
+    )
+
+    for match in re_detail.finditer(stdout):
+        if match.group('uuid') in mdarray_list:
+            mdarray_list[match.group('uuid')].active = True
+            mdarray_list[match.group('uuid')].refresh_state()
+            if match.group('spares'):
+                mdarray_list[match.group('uuid')].sparses = match.group('spares')
+
+    return mdarray_list
+
+
+def check_and_get_devices(devices, ansmod):
+    checked_devices = []
+    for device in devices:
+        if not os.path.exists(device):
+            ansmod.fail_json(msg="%s does not exist" % device)
+        if os.path.islink(device):
+            checked_devices.append(os.path.realpath(device))
+        else:
+            checked_devices.append(device)
+    return checked_devices
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str', required=True),
+            devices=dict(type='list', default=[]),
+            raid_devices=dict(type='int'),
+            level=dict(type='str', default='raid10', choices=[
+                'linear',
+                'stripe',
+                'raid0',
+                'raid1',
+                'raid4',
+                'raid5',
+                'raid6',
+                'raid10',
+                'multipath',
+                'faulty',
+                'container'
+            ]),
+            metadata=dict(type='str', default='1.2', choices=['0', '0.90', '1.0', '1.1', '1.2']),
+            opts=dict(type='str'),
+            state=dict(type='str', default='present', choices=['absent', 'present', 'info']),
+        ),
+        supports_check_mode=False,
```

Comment:
out of curiosity, is it feasible to support check mode?

Reply:
@maxamillion it's too risky because even to just read the status of raid array i have to `assemble` the array and this step can start a job of rebuild for example which actually can do read/write on disk

Reply:
In that case, let's make sure to identify in the documentation section very explicitly that this module isn't idempotent.